service/worker/deletenamespace/deleteexecutions/activities.go

```diff
@@ -0,0 +1,160 @@
+		return nil, err
+	}
+
+	return resp.NextPageToken, nil
+}
+
+func (a *Activities) DeleteExecutionsActivity(ctx context.Context, params DeleteExecutionsActivityParams) (DeleteExecutionsActivityResult, error) {
+	rateLimiter := quotas.NewRateLimiter(float64(params.DeleteRPS), params.DeleteRPS)
+
+	var result DeleteExecutionsActivityResult
+
+	req := &manager.ListWorkflowExecutionsRequestV2{
+		NamespaceID:   params.NamespaceID,
+		Namespace:     params.Namespace,
+		PageSize:      params.ListPageSize,
+		NextPageToken: params.NextPageToken,
+	}
+	resp, err := a.visibilityManager.ListWorkflowExecutions(req)
+	if err != nil {
+		a.metricsClient.IncCounter(metrics.DeleteExecutionsWorkflowScope, metrics.ListExecutionsFailuresCount)
+		a.logger.Error("Unable to list all workflows.", tag.WorkflowNamespace(params.Namespace.String()), tag.Error(err))
+		return result, err
+	}
+	for _, execution := range resp.Executions {
+		_ = rateLimiter.Wait(ctx)
+		if execution.Status == enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING {
+			_, err := a.historyClient.TerminateWorkflowExecution(ctx, &historyservice.TerminateWorkflowExecutionRequest{
+				NamespaceId: params.NamespaceID.String(),
+				TerminateRequest: &workflowservice.TerminateWorkflowExecutionRequest{
+					Namespace:         params.Namespace.String(),
+					WorkflowExecution: execution.Execution,
+					Reason:            "Delete namespace",
+				},
+			})
+			switch err.(type) {
+			case nil, *serviceerror.NotFound: // Workflow already completed or doesn't exist.
+			default:
+				a.metricsClient.IncCounter(metrics.DeleteExecutionsWorkflowScope, metrics.TerminateExecutionFailuresCount)
+				a.logger.Error("Unable to terminate workflow execution.", tag.WorkflowNamespace(params.Namespace.String()), tag.WorkflowID(execution.Execution.GetWorkflowId()), tag.WorkflowRunID(execution.Execution.GetRunId()), tag.Error(err))
+				result.ErrorCount++
+				continue
+			}
+		}
+		_, err = a.historyClient.DeleteWorkflowExecution(ctx, &historyservice.DeleteWorkflowExecutionRequest{
+			NamespaceId:       params.NamespaceID.String(),
+			WorkflowExecution: execution.Execution,
+		})
+		switch err.(type) {
+		case nil, *serviceerror.NotFound: // Workflow doesn't exist. Treat it as success.
+			result.SuccessCount++
+			a.metricsClient.IncCounter(metrics.DeleteExecutionsWorkflowScope, metrics.DeleteExecutionsSuccessCount)
+		default:
+			result.ErrorCount++
+			a.metricsClient.IncCounter(metrics.DeleteExecutionsWorkflowScope, metrics.DeleteExecutionFailuresCount)
+			a.logger.Error("Unable to delete workflow execution.", tag.WorkflowNamespace(params.Namespace.String()), tag.WorkflowID(execution.Execution.GetWorkflowId()), tag.WorkflowRunID(execution.Execution.GetRunId()), tag.Error(err))
+		}
+		if (result.SuccessCount+result.ErrorCount)%heartbeatEveryExecutions == 0 {
+			activity.RecordHeartbeat(ctx, result)
```

Comment:
heartbeat should report progress (the index of the processed execution in the list), and the beginning of this activity should recover from heartbeat if there is any.

Reply:
If the list result is 1000, and activity failed after 500 executions processed, we should retry from 501 executions.